feat: 실시간 데이터 처리 스파크 애플리케이션 추가

Ra99it opened the pull request:
### 변경 사항 개요
- Spark Streaming으로 Kafka를 읽어 데이터 중 개인을 식별할 수 있는 유저명을 SHA-256를 적용에 암호화(가명화) 작업을 하고 최종적으로 S3에 저장합니다

### 관련 이슈
  - #17 

### 변경 사항 상세 설명
1. Zeppelin 환경에서 작성했던 코드를 spark-submit 할 수 있게 작성하였습니다.
2. 개인을 식별 할 수 있는 유저명을 SHA-256을 적용 해 데이터 암호화(가명화)를 적용했습니다. 이로써 접근한 사용자는 식별할 수 없게 됩니다.
3. 기존은 txt파일로 저장했지만, 현재는 json 파일 형식으로 저장을 진행합니다. 텍스트로 저장하면 데이터를 사용할 때 추가 변환 작업을 요구하게 됩니다. 데이터의 품질이 낮아진다 판단해 json 형식으로 저장하도록 변경하였습니다.

### 테스트
- 직접 수동으로 테스트를 진행했습니다. 직접 클러스터에 spark-submit으로 스파크 어플리케이션을 제출 후, 구조적 스트리밍이 작동하면서 최종적으로 암호화 된 데이터를 저장하는 것 까지 완료하였습니다.

**실행 명령**
> spark-submit  \
--master yarn \
 --deploy-mode cluster \
--conf spark.kafka.bootstrap.servers="airflow-mysql-01:29092" \
--conf spark.kafka.topic="lol" \ 
--packages org.apache.spark:spark-sql-kafka-0-10_2.12:3.3.0,org.apache.spark:spark-streaming-kafka-0-10_2.12:3.3.0 \ 
./spark_streaming_kafka_to_s3.py

### 추가 고려 사항
- 소스에 옵션으로 spark.jars.packages를 작성했지만, spark-submit할 때 인식하지 못하는 것 같습니다. 그래서 제출할 때 패키지의 정보도 같이 입력해야 합니다.

### 스크린샷 (필요시)
- 변경된 UI가 있는 경우, 전후의 스크린샷을 첨부하세요.
- 변경 사항을 시각적으로 보여줄 수 있는 자료를 제공하세요.

### 체크리스트
- [X] 코드가 정상적으로 컴파일되고 빌드됩니다.
- [ ] 모든 테스트가 통과되었습니다.
- [ ] 관련 문서가 업데이트되었습니다.
- [X] 팀원들이 리뷰할 준비가 되었습니다.

### 기타
- 추가로 공유하고 싶은 내용이나 문의 사항이 있으면 작성하세요.

Ra99it committed:
feat: 실시간 데이터 처리 스파크 애플리케이션 추가

Spark Streaming으로 Kafka를 읽어 데이터 중 개인을 식별할 수 있는 유저명을 SHA-256를 적용에 암호화(익명) 작업을 하고 최종적으로 S3에 저장합니다

Ra99it committed:
Merge branch 'develop' of https://github.com/f-lab-edu/league-of-legends-data-solution into feature/#17

f-lab-johnp commented on Spark_Application/spark_streaming_kafka_to_s3.py:
```diff
@@ -0,0 +1,62 @@
+  .select(from_json(col("value"), schema).alias("data")) \
+  .select("data.*")
+
+jsonParsedStream = jsonParsedStream.withColumn("account_sha2", sha2(col("account"), 256))
```
`account_sha2` 라는 이름은 암호화 방식을 유추할 수 있을 것 같습니다.
보안을 위해 다른 이름을 사용하는게 어떨까요? ex. `account_no`

f-lab-johnp commented on Spark_Application/spark_streaming_kafka_to_s3.py:
```diff
@@ -0,0 +1,62 @@
+  .trigger(processingTime='1 minute') \
+  .outputMode("append") \
+  .format("json") \
+  .option("path", "s3://sjm-simple-data/bronzelayer/playerlogs/") \
```
S3에서 테이블 경로를 어떻게 설정 했는지 설명해주실 수 있나요?
현재 작성된 경로라면 `DB: bronzelayer`, `테이블: playerlogs` 가 맞을까요?

Ra99it replied:
원시 데이터를 저장하는 Bronze Layer 역할을 하는 S3 객체 스토리지입니다.

현재의 Bronze Layer는 최소 변환을 거친 원본 데이터만 저장하고 있기 때문에 데이터 레이크와 비슷한 역할을 하고 있습니다. 그래서 DB, 테이블이라기 보다는 디렉토리 경로입니다.

앞으로 계획은 Delta Lake를 사용해 Delta Lake의 테이블로 관리할 예정입니다. 현재는 단순히 해당 경로에 변환된 데이터를 저장하고 있는 상황입니다.

f-lab-johnp replied:
결국 S3 디렉토리 경로가 메타스토어에 테이블 경로로 매핑되는게 아닐까요?
데이터레이크에서는 여러 개의 스키마, 테이블을 운영할 수 있어야 합니다.
그리고 S3는 경로 prefix에 따라 요청 제한이 존재합니다. ([S3 문서](https://docs.aws.amazon.com/ko_kr/AmazonS3/latest/userguide/optimizing-performance.html))
따라서 모든 raw 데이터를 bronzelayer prefix 하위에 저장하는 것보다 구분할 수 있도록 수정하는게 어떨까요?
```
경로 규칙: 버킷 / 레이어_스키마명 / 테이블명 / 파티션
예시: s3://sjm-simple-data/bronze_lol/playerlogs/dt=2024-09-01/ *.json
```

Ra99it replied:
아 결국엔 디렉토리 경로가 곧 매핑이 되어서 여러 개의 스키마, 테이블로 운영이 되어야 한다는 거군요. 이 사실을 모르고 있었던 거 같습니다. 추가적으로 더 알아보도록 하겠습니다.

그런 관점을 알고 소스를 보면 너무 일반 디렉토리에 데이터를 저장하는 것처럼 느껴지네요. 이렇게 된다면 Trino나 다른 쿼리 엔진으로 데이터를 접근 할 때 일관성이 없어서 불편하고, 효율적이지 않겠군요.

현재 문제점을 인지하였고 다음과 같이 변경하였습니다.
s3://sjm-simple-data/bronze_riot/playerlogs/create_room_date=2024-09-01/*.json 으로 변경되었습니다.
**create_room_date**을 기준으로 파티션을 나누도록 했습니다.

추가적으로 S3가 요청 제한이 있다는 걸 새로 알았습니다! 데이터가 너무 많아질 시 파티션 기준을 create_room_date과 추가로 room_id와 같은 계획을 세워보도록 하겠습니다.

f-lab-johnp commented on Spark_Application/spark_streaming_kafka_to_s3.py:
```diff
@@ -0,0 +1,62 @@
+jsonParsedStream = jsonParsedStream.withColumn("account_sha2", sha2(col("account"), 256))
+jsonParsedStream = jsonParsedStream.drop("account")
+
+transformedPlayerLogs = jsonParsedStream.toDF("createRoomDate", "method", "ingametime", "ip",
+                           "deathCount", "roomID", "datetime", "x", "y",
+                           "inputkey", "champion", "status", "account_sha2")
```
작성하신 컬럼명에 대한 convention이 혼재 되어 있는 것 같습니다.
사용하는 카탈로그에 따라 대소문자를 구분을 지원하지 않는 경우도 있습니다.
그리고 오류 최소화와 일관된 정책을 위해 일반적으로 대소문자를 구분하지 않는 것(`case-insensitive`)이 권장됩니다.

Ra99it committed:
refactor: 편의성을 위한 코드 변경 (Kafka 서버 외부 입력, S3 저장 시 날짜를 기준)

Kafka 서버는 보안 상 외부에서 입력하도록 변경하였습니다. 그리고 데이터가 저장되는 위치와 checkpoint의 가독성을 위해서 오늘 날자를 기준 디렉토리를 생성하고 그 안에 데이터를 생성합니다. 또한 코드 컨벤션을 적용하였고, 유저명 뿐만 아니라 IP 정보도 수집하지 않도록 변경하였습니다.

Ra99it commented:
## 변경사항
1. Kafka 서버와 토픽 정보를 외부에서 입력하도록 변경하였습니다.
2. 데이터의 컬럼명을 소문자로 통일했습니다.
3. IP는 수집하지 않도록 했습니다.
4. 데이터가 저장되는 위치를 찾기 쉽도록 오늘 날짜를  기준으로 디렉토리안에 데이터가 생성됩니다.

Ra99it committed:
refactor: writeStream.path를 구분있게 구조적으로 변경 및 create_room_date을 기준으로 Partition 적용
